File: Java/FileSystem/Folder.java
import java.util.ArrayList;
import java.util.List;

public class Folder extends AbstractFileSystemNode {

    private final List<AbstractFileSystemNode> childs = new ArrayList<>();

    public Folder(String inName, AbstractFileSystemNode inParent, AbstractFileSystemNode... inNodes) {
        setName(inName);
        if (inParent instanceof Folder) {
            setParent(inParent);
            Folder folderParent = (Folder) inParent;
            folderParent.getChilds().add(this);
        } else {
            setParent(null);
        }
        for (AbstractFileSystemNode node : inNodes) {
            childs.add(node);
            if (node instanceof Folder) {
                Folder parentNode = (Folder) node.getParent();
                parentNode.getChilds().remove(node);
            }
            node.setParent(this);
        }
    }

    @Override
    public String getName() {
        return name + "/";
    }

    public List<AbstractFileSystemNode> getChilds() {
        return childs;
    }

}

File: Kotlin/FileSystemOnKotlin/src/main/kotlin/Folder.kt
class Folder(val childs: MutableList<AbstractFileSystemNode> = mutableListOf()) : AbstractFileSystemNode() {
    override var name: String = ""
        get() = "$field/"

    constructor(
        inName: String,
        inParent: AbstractFileSystemNode?,
        vararg inNodes: AbstractFileSystemNode
    ) : this() {
        name = inName
        if (inParent is Folder) {
            parent = inParent
            inParent.childs.add(this)
        }
        for (node in inNodes) {
            childs.add(node)
            if (node is Folder) {
                val parentNode = node.parent as Folder
                parentNode.childs.remove(node)
            }
            node.parent = this
        }
    }
}
